backend-rs/src/utils.rs

```diff
@@ -0,0 +1,425 @@
+// Heavily inspired by https://github.com/D1plo1d/h264_webcam_stream
+impl<'a> H264CameraReader<'a> {
+    pub fn new(device: &mut Device, mode: CameraMode) -> Result<Self, Box<dyn Error>> {
+        let format = Format::new(mode.width, mode.height, FourCC::new(b"MJPG"));
+        Capture::set_format(device, &format)?;
+
+        let parameters = Parameters::new(mode.frame_interval);
+        Capture::set_params(device, &parameters)?;
+
+        let stream = MmapStream::with_buffers(device, Type::VideoCapture, 4)?;
+        let encoder = Encoder::new()?;
+
+        Ok(Self {
+            stream,
+            camera_mode: mode,
+            h264_encoder: encoder,
+        })
+    }
+
+    pub fn read(&mut self) -> Result<Vec<u8>, Box<dyn Error>> {
+        let (buffer, _) = self.stream.next()?;
+        let mut jpg = Decoder::new(buffer);
+        let pixels = jpg.decode()?;
+
+        let yuv_buffer = YUVBuffer::from_rgb_source(RgbSliceU8::new(
+            &pixels[..],
+            (
+                self.camera_mode.width as usize,
+                self.camera_mode.height as usize,
+            ),
+        ));
+        let bitstream = self.h264_encoder.encode(&yuv_buffer)?;
+
+        Ok(bitstream.to_vec())
+    }
+}
+
+// CameraMode(s) are used for controlling what resolution and frame rate the H264CameraReader operates at.
+#[derive(Debug, Clone, Copy)]
+pub struct CameraMode {
+    width: u32,
+    height: u32,
+    frame_interval: Fraction, // 1/30 is 30 fps, 1/15 is 15fps, etc.
+}
+
+impl ToString for CameraMode {
+    fn to_string(&self) -> String {
+        format!(
+            "{}x{} @{}fps",
+            self.width, self.height, self.frame_interval.denominator
+        )
+    }
+}
+
+impl CameraMode {
+    // Fetch all possible CameraMode(s) of the Device camera.
+    pub fn fetch_all(device: &Device) -> Result<Vec<CameraMode>, Box<dyn Error>> {
+        let mut configs: Vec<CameraMode> = Vec::new();
+
+        let discrete_iter = Capture::enum_framesizes(device, FourCC::new(b"MJPG"))?
+            .into_iter()
+            .flat_map(|framesize| framesize.size.to_discrete());
+
+        for discrete_frame_size in discrete_iter {
+            for interval in Capture::enum_frameintervals(
+                device,
+                FourCC::new(b"MJPG"),
+                discrete_frame_size.width,
+                discrete_frame_size.height,
+            )? {
+                let fraction = match interval.interval {
+                    FrameIntervalEnum::Discrete(fraction) => fraction,
+                    FrameIntervalEnum::Stepwise(Stepwise { min, .. }) => min,
+                };
+
+                configs.push(CameraMode {
+                    width: discrete_frame_size.width,
+                    height: discrete_frame_size.height,
+                    frame_interval: fraction,
+                });
+            }
+        }
+
+        Ok(configs)
+    }
+}
+
+// Handles communication between CameraThreadHandle(s) and WebRTC clients.
```

Comment:
You may have intended to use a doc comment here: 
```suggestion
/// Handles communication between CameraThreadHandle(s) and WebRTC clients.
```